python/cufinufft/cufinufft/_plan.py

```diff
@@ -0,0 +1,373 @@
+#!/usr/bin/env python
+"""
+This module contains the high level python wrapper for
+the cufinufft CUDA libraries.
+"""
+
+import atexit
+import sys
+import warnings
+
+import numpy as np
+
+from ctypes import byref
+from ctypes import c_int
+from ctypes import c_void_p
+
+from cufinufft._cufinufft import NufftOpts
+from cufinufft._cufinufft import _default_opts
+from cufinufft._cufinufft import _make_plan
+from cufinufft._cufinufft import _make_planf
+from cufinufft._cufinufft import _set_pts
+from cufinufft._cufinufft import _set_ptsf
+from cufinufft._cufinufft import _exec_plan
+from cufinufft._cufinufft import _exec_planf
+from cufinufft._cufinufft import _destroy_plan
+from cufinufft._cufinufft import _destroy_planf
+
+from pycuda.gpuarray import GPUArray
+
+
+# If we are shutting down python, we don't need to run __del__
+#   This will avoid any shutdown gc ordering problems.
+exiting = False
+atexit.register(setattr, sys.modules[__name__], 'exiting', True)
+
+
+class Plan:
+    """
+    A non-uniform fast Fourier transform (NUFFT) plan
+
+    The ``Plan`` class lets the user exercise more fine-grained control over
+    the execution of an NUFFT. First, the plan is created with a certain set
+    of parameters (type, mode configuration, tolerance, sign, number of
+    simultaneous transforms, and so on). Then the nonuniform points are set
+    (source or target depending on the type). Finally, the plan is executed on
+    some data, yielding the desired output.
+
+    Args:
+        nufft_type      (int): type of NUFFT (1 or 2).
+        n_modes         (tuple of ints): the number of modes in each
+                        dimension (for example `(50, 100)`).
+        n_trans         (int, optional): number of transforms to compute.
+        eps             (float, optional): precision requested (>1e-16).
+        isign           (int, optional): if +1, uses the positive sign
+                        exponential, otherwise the negative sign; defaults to
+                        +1 for type 1 and to -1 for type 2.
+        dtype           (string, optional): the precision of the transofrm,
+                        'complex64' or 'complex128'.
+        **kwargs        (optional): additional options corresponding to the
+                        entries in the `nufft_opts` structure may be specified
+                        as keyword-only arguments.
+    """
+
+    def __init__(self, nufft_type, n_modes, n_trans=1, eps=1e-6, isign=None,
+                 dtype="complex64", **kwargs):
+        if isign is None:
```

Comment:
Can we add type signatures, at least for the public interfaces? I'd be happy to do a later PR with them regardless.

Reply:
I'd prefer to do this in a separate PR (haven't worked with types in Python yet, so this will be a bit of a learning experience).